util/eventbus/eventbustest/eventbustest.go

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package eventbustest
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"tailscale.com/util/eventbus"
+)
+
+// NewBus constructs an [eventbus.Bus] that will be shut automatically when
+// its controlling test ends.
+func NewBus(t *testing.T) *eventbus.Bus {
+	bus := eventbus.New()
+	t.Cleanup(bus.Close)
+	return bus
+}
+
+// NewTestWatcher constructs a [Watcher] that can be used to check the stream of
+// events generated by code under test. After construction the caller may use
+// [Expect], [Expect] and [ExpectExactly], to verify that the desired
+// events were captured.
+func NewWatcher(t *testing.T, bus *eventbus.Bus) *Watcher {
+	tw := &Watcher{
+		mon:     bus.Debugger().WatchBus(),
+		TimeOut: 5 * time.Second,
+		chDone:  make(chan bool, 1),
+		events:  make(chan any, 100),
+	}
+	if deadline, ok := t.Deadline(); ok {
+		tw.TimeOut = deadline.Sub(time.Now())
+	}
+	t.Cleanup(tw.done)
+	go tw.watch()
+	return tw
+}
+
+// Watcher monitors and holds events for test expectations.
+type Watcher struct {
+	mon    *eventbus.Subscriber[eventbus.RoutedEvent]
+	events chan any
+	chDone chan bool
+	// TimeOut defines when the Expect* functions should stop looking for events
+	// coming from the Watcher. The value is set by [NewWatcher] and defaults to
+	// the deadline passed in by [testing.T]. If looking to verify the absence
+	// of an event, the TimeOut can be set to a lower value after creating the
+	// Watcher.
+	TimeOut time.Duration
+}
+
+// Type is a helper representing the expectation to see an event of type T, without
+// caring about the content of the event.
+// It makes it possible to use helpers like:
+//
+//	eventbustest.ExpectFilter(tw, eventbustest.Type[EventFoo]())
+func Type[T any]() func(T) { return func(T) {} }
+
+// Expect verifies that the given events are a subsequence of the events
+// observed by tw. That is, tw must contain at least one event matching the type
+// of each argument in the given order, other event types are allowed to occur in
+// between without error. The given events are represented by a function
+// that can the the form of:
```

Comment:
typo: the the
```suggestion
// that must have one of the following forms:
```
maybe?